src/util.h

```diff
@@ -112,6 +108,64 @@ static inline bool error(const char* format)
     return false;
 }
 
+/* One of these structures is statically allocated at the site of every call to
+ * LogPrint(category, format, ...), and also at the site of every call to
+ * LogAcceptCategory(category). The first time a site is reached, its structure
+ * is initialized and linked into the global linked list logFastReached (this
+ * "discovery" event is the first time it's possible to do this).
+ */
+struct LogFastEntry {
+    const char* category;
+    const char* line;   // pathname:lineNumber
+    bool enabled;
+    LogFastEntry *next;
+};
+
+/* This function is called the first time (and only the first time) a call
+ * site to LogPrint() or LogAcceptCategory() is visited.
+ */
+void LogFastEntryInit(LogFastEntry *lfe, const char *category, const char* line);
+
+/**
+ * Print the given string (with arguments) if the category is enabled, or if
+ * the line (where the LogPrint occurs) is enabled. This is very efficient
+ * except possibly for the first call in the life of the process.
+ */
+#define LogPrint(category, ...) do {                            \
+    static_assert((category) != nullptr);                       \
```

Comment:
This assert also guarantees that `category` is a constant (at compile-time) string (see below for the couple of places I had to change where that wasn't the case).